File: Projects/learn-spring-framework-01/src/main/java/com/feeeeel/learnspringframework/helloworld/HelloWorldConfiguration.java
package com.feeeeel.learnspringframework.helloworld;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Primary;

//Eliminate verbosity in creating Java Beans
//Public accessor methods, constructor,
//equals, hashCode and toString are automatically created.
//Released in JDK 16.

record Person(String name, int age, Address address) {};

// Address - firstLine & city
record Address(String firstLine, String city) {}
@Configuration
public class HelloWorldConfiguration {

    @Bean
    public String name() {
        return "Feeeeel";
    }

    @Bean
    public int age() {
        return 25;
    }

    @Bean
    public Person person() {
        return new Person("Ravi", 20, new Address("Gwangjin", "Seoul"));
    }

    @Bean
    public Person person2MethodCall() {
        return new Person(name(), age(), address()); // name, age, address
    }

    @Bean
    public Person person3Parameters(String name, int age, Address address3) { // name, age, address3
        return new Person(name, age, address3);
    }

    @Bean
    @Primary
    public Person person4Parameters(String name, int age, Address address) { // name, age, address3
        return new Person(name, age, address);
    }

    @Bean
    public Person person5Qualifier(String name, int age, @Qualifier("address3qualifier") Address address) { // name, age, address3
        return new Person(name, age, address);
    }

    @Bean(name = "address2")
    @Primary
    public Address address() {
        return new Address("Nowon", "Seoul");
    }

    @Bean(name = "address3")
    @Qualifier("address3qualifier")
    public Address address3() {
        return new Address("Baker Street", "London");
    }
}

File: Projects/learn-spring-framwork-02-kotlin/src/main/kotlin/com/feeeeel/learnspringframwork01kotlin/helloworld/HelloWorldConfiguration.kt
package com.feeeeel.learnspringframwork01kotlin.helloworld

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Primary

data class Person(val name: String, val age: Int, val address: Address)
data class Address(val firstLine: String, val city: String)

@Configuration
class HelloWorldConfiguration {

    @Bean
    fun name(): String = "Feeeeel"

    @Bean
    fun age(): Int = 25

    @Bean
    fun person(): Person = Person("Feeeeel", 25, Address("Nowon", "Seoul"))

    @Bean
    fun person2MethodCall(): Person = Person(name(), age(), address())

    @Bean
    fun person3Parameters(name: String, age: Int, address3: Address): Person = Person(name, age, address3)

    @Bean
    @Primary
    fun person4Parameters(name: String, age: Int, address: Address): Person = Person(name, age, address)

    @Bean(name = ["address2"])
    @Primary
    fun address(): Address = Address("Nowon", "Seoul")

    @Bean(name = ["address3"])
    fun address3(): Address = Address("Shibuya", "Tokyo")

}